File: ios/Sailings/APIClient.swift
import Foundation

enum APIError: LocalizedError {
    case unauthorized
    case message(String)
    case validation([String])
    case transport(Error)
    case decoding(Error)

    var errorDescription: String? {
        switch self {
        case .unauthorized:
            return "Your session has expired. Please sign in again."
        case .message(let text):
            return text
        case .validation(let errors):
            return errors.joined(separator: "\n")
        case .transport:
            return "Couldn't reach the server. Check your connection and that the server is running."
        case .decoding:
            return "The server sent something unexpected."
        }
    }
}

actor APIClient {
    static let shared = APIClient()

    private let session = URLSession(configuration: .default)
    private var token: String?

    func setToken(_ token: String?) {
        self.token = token
    }

    // MARK: - Endpoints

    func login(emailAddress: String, password: String) async throws -> LoginResponse {
        try await send(.post, "/session", body: ["email_address": emailAddress, "password": password])
    }

    func logout() async throws {
        _ = try await raw(.delete, "/session", body: nil)
    }

    func profile() async throws -> Profile {
        try await send(.get, "/profile")
    }

    func voyages() async throws -> [Voyage] {
        try await send(.get, "/sailings")
    }

    func voyage(id: Int) async throws -> Voyage {
        try await send(.get, "/sailings/\(id)")
    }

    func registrations() async throws -> [Registration] {
        try await send(.get, "/registrations")
    }

    /// `climbing` follows the Rails encoding: 1 = Yes, 2 = No.
    func register(voyageID: Int, comment: String?, climbing: Int?) async throws -> Registration {
        var body: [String: Any] = [:]
        if let comment, !comment.isEmpty { body["comment"] = comment }
        if let climbing { body["climbing"] = climbing }
        return try await send(.post, "/sailings/\(voyageID)/registrations", body: body)
    }

    func cancel(registrationID: Int) async throws {
        _ = try await raw(.delete, "/registrations/\(registrationID)", body: nil)
    }

    // MARK: - Plumbing

    private enum Method: String {
        case get = "GET", post = "POST", delete = "DELETE"
    }

    private func send<T: Decodable>(_ method: Method, _ path: String, body: [String: Any]? = nil) async throws -> T {
        let data = try await raw(method, path, body: body)
        do {
            return try Self.decoder.decode(T.self, from: data)
        } catch {
            throw APIError.decoding(error)
        }
    }

    private func raw(_ method: Method, _ path: String, body: [String: Any]?) async throws -> Data {
        guard let url = URL(string: Config.apiRoot + path) else {
            throw APIError.message("Bad URL: \(path)")
        }

        var request = URLRequest(url: url)
        request.httpMethod = method.rawValue
        request.setValue("application/json", forHTTPHeaderField: "Accept")
        if let token {
            request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }
        if let body {
            request.setValue("application/json", forHTTPHeaderField: "Content-Type")
            request.httpBody = try JSONSerialization.data(withJSONObject: body)
        }

        let data: Data
        let response: URLResponse
        do {
            (data, response) = try await session.data(for: request)
        } catch {
            throw APIError.transport(error)
        }

        guard let http = response as? HTTPURLResponse else {
            throw APIError.message("Unexpected response from the server.")
        }

        switch http.statusCode {
        case 200..<300:
            return data
        case 401:
            throw APIError.unauthorized
        case 422:
            throw APIError.validation(Self.validationErrors(in: data) ?? ["That didn't work."])
        default:
            throw APIError.message(Self.errorMessage(in: data) ?? "Request failed (HTTP \(http.statusCode)).")
        }
    }

    /// Rails sends `{"error": "..."}` for single failures and `{"errors": [...]}` for validation.
    private static func errorMessage(in data: Data) -> String? {
        struct Payload: Decodable { let error: String }
        return try? JSONDecoder().decode(Payload.self, from: data).error
    }

    private static func validationErrors(in data: Data) -> [String]? {
        struct Payload: Decodable { let errors: [String] }
        return try? JSONDecoder().decode(Payload.self, from: data).errors
    }

    /// The API mixes two date shapes: full ISO-8601 timestamps (departs_at) and
    /// bare calendar dates (expires_on, last_sailed). A single built-in strategy
    /// handles only one, so try both.
    ///
    /// These are parse *strategies* (value types) rather than DateFormatter
    /// instances on purpose: JSONDecoder may call this closure concurrently, and
    /// formatters are reference types that are not thread-safe.
    private static let decoder: JSONDecoder = {
        let decoder = JSONDecoder()
        decoder.keyDecodingStrategy = .convertFromSnakeCase
        decoder.dateDecodingStrategy = .custom { decoder in
            let text = try decoder.singleValueContainer().decode(String.self)
            let dateOnly = Date.ISO8601FormatStyle(timeZone: .gmt).year().month().day()

            if let date = try? Date(text, strategy: .iso8601) { return date }
            if let date = try? Date(text, strategy: dateOnly) { return date }
            throw DecodingError.dataCorrupted(
                .init(codingPath: decoder.codingPath, debugDescription: "Unrecognised date: \(text)")
            )
        }
        return decoder
    }()
}

File: android/app/src/main/java/org/sailings/app/ApiClient.kt
package org.sailings.app

import io.ktor.client.HttpClient
import io.ktor.client.engine.okhttp.OkHttp
import io.ktor.client.request.header
import io.ktor.client.request.request
import io.ktor.client.request.setBody
import io.ktor.client.statement.bodyAsText
import io.ktor.http.ContentType
import io.ktor.http.HttpHeaders
import io.ktor.http.HttpMethod
import io.ktor.http.contentType
import kotlinx.serialization.ExperimentalSerializationApi
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonNamingStrategy
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.put

/** Mirrors the iOS APIError: a typed failure the UI can render directly. */
sealed class ApiException(message: String) : Exception(message) {
    /** The token is gone (expired, logged out elsewhere, account disabled). */
    data object Unauthorized :
        ApiException("Your session has expired. Please sign in again.")

    /** A single server-supplied message, from `{"error": "..."}`. */
    class Message(val text: String) : ApiException(text)

    /** Validation failures, from `{"errors": [...]}`. */
    class Validation(val errors: List<String>) : ApiException(errors.joinToString("\n"))

    /** Couldn't reach the server at all. */
    class Transport(cause: Throwable) :
        ApiException("Couldn't reach the server. Check your connection and that the server is running.") {
        init { initCause(cause) }
    }

    /** The server replied, but the body didn't decode. */
    class Decoding(cause: Throwable) :
        ApiException("The server sent something unexpected.") {
        init { initCause(cause) }
    }
}

@OptIn(ExperimentalSerializationApi::class)
class ApiClient(private val baseUrl: String = Config.API_ROOT) {

    /** snake_case keys, tolerant of unknown fields, nulls omitted when encoding. */
    private val json = Json {
        namingStrategy = JsonNamingStrategy.SnakeCase
        ignoreUnknownKeys = true
        explicitNulls = false
    }

    private val http = HttpClient(OkHttp) {
        expectSuccess = false
    }

    /** Set after login, cleared on logout or any 401. */
    @Volatile
    var token: String? = null

    // region Endpoints

    suspend fun login(emailAddress: String, password: String): LoginResponse {
        val body = buildJsonObject {
            put("email_address", emailAddress)
            put("password", password)
        }
        return decode(request(HttpMethod.Post, "/session", body))
    }

    suspend fun logout() {
        request(HttpMethod.Delete, "/session")
    }

    suspend fun profile(): Profile =
        decode(request(HttpMethod.Get, "/profile"))

    suspend fun voyages(): List<Voyage> =
        decode(request(HttpMethod.Get, "/sailings"))

    suspend fun voyage(id: Int): Voyage =
        decode(request(HttpMethod.Get, "/sailings/$id"))

    suspend fun registrations(): List<Registration> =
        decode(request(HttpMethod.Get, "/registrations"))

    /** `climbing` follows the Rails encoding: 1 = Yes, 2 = No. */
    suspend fun register(voyageId: Int, comment: String?, climbing: Int?): Registration {
        val body = buildJsonObject {
            if (!comment.isNullOrEmpty()) put("comment", comment)
            if (climbing != null) put("climbing", climbing)
        }
        return decode(request(HttpMethod.Post, "/sailings/$voyageId/registrations", body))
    }

    suspend fun cancel(registrationId: Int) {
        request(HttpMethod.Delete, "/registrations/$registrationId")
    }

    // endregion

    // region Plumbing

    private inline fun <reified T> decode(text: String): T =
        try {
            json.decodeFromString<T>(text)
        } catch (e: Exception) {
            throw ApiException.Decoding(e)
        }

    /** Performs the request and maps the HTTP status to a typed result or exception. */
    private suspend fun request(
        method: HttpMethod,
        path: String,
        body: JsonObject? = null,
    ): String {
        val response = try {
            http.request(baseUrl + path) {
                this.method = method
                header(HttpHeaders.Accept, ContentType.Application.Json.toString())
                token?.let { header(HttpHeaders.Authorization, "Bearer $it") }
                if (body != null) {
                    contentType(ContentType.Application.Json)
                    setBody(json.encodeToString(JsonObject.serializer(), body))
                }
            }
        } catch (e: Exception) {
            android.util.Log.e("ApiClient", "transport failure: $method ${baseUrl + path}", e)
            throw ApiException.Transport(e)
        }

        val text = response.bodyAsText()
        return when (response.status.value) {
            in 200..299 -> text
            401 -> throw ApiException.Unauthorized
            422 -> throw ApiException.Validation(validationErrors(text) ?: listOf("That didn't work."))
            else -> throw ApiException.Message(errorMessage(text) ?: "Request failed (HTTP ${response.status.value}).")
        }
    }

    /** Rails sends `{"error": "..."}` for single failures. */
    private fun errorMessage(text: String): String? =
        runCatching { (json.parseToJsonElement(text) as? JsonObject)?.get("error") }
            .getOrNull()
            .let { (it as? JsonPrimitive)?.content }

    /** ...and `{"errors": [...]}` for validation. */
    private fun validationErrors(text: String): List<String>? =
        runCatching {
            (json.parseToJsonElement(text) as? JsonObject)?.get("errors")
        }.getOrNull()
            ?.let { it as? kotlinx.serialization.json.JsonArray }
            ?.mapNotNull { (it as? JsonPrimitive)?.content }

    // endregion
}
